[chore] Add assign-codeowners workflow

TylerHelmuth opened the pull request:
**Description:**
Adds workflow, similar to ping-codeowners, that will assign code owners to a PR based on a component label.

gh expects the `--add-assignee` flag to be the exact github username; `@` cannot be included.  Comma separated names is supported.

**Testing:**
https://github.com/TylerHelmuth/opentelemetry-collector-contrib/pull/10

TylerHelmuth committed:
Add assign-codeowners workflow

codeboten reviewed (commented):
Nice! I was looking at this earlier this week as well. A couple of open questions:

* What happens if the code owner is the original author of the PR? i'd guess we wouldn't want them to be assigned to it
* In the event that a valid component is labeled, should we remove the approver that's already assigned?

codeboten reviewed (commented):
One more question, would it make sense to add a comment in the PR to mention the reason for assignment? 

This makes me think assignment to code owners should be done for issues as well, rather than just mentioning them in the issue.

TylerHelmuth commented:
> What happens if the code owner is the original author of the PR? i'd guess we wouldn't want them to be assigned to it

Good question.  At the moment it would definitely attempt to assign them, but not sure what the outcome would be.  I will test this.  

> In the event that a valid component is labeled, should we remove the approver that's already assigned?

I think eventually we could replace the existing auto-assign workflow that is using the approvers/maintainers list, but only once we've ensured that all components have Code Owners that are OpenTelemetry members AND all components have an approver/maintainer sponsor.  I like that the existing workflow guarantees an assignment of someone with Approve rights.

>  would it make sense to add a comment in the PR to mention the reason for assignment

Definitely.  I will add this functionality.

> This makes me think assignment to code owners should be done for issues as well, rather than just mentioning them in the issue.

I kinda like pinging still for issues, since it feels like we leave issues unassigned at the moment until someone is actively working on it.  @evan-bradley what do you think?

TylerHelmuth commented:
I also just noticed that when the workflow fails it'll show up in the PR's checks.  I think I like that, as it'll identify scenarios where code owners are not OpenTelemetry members, which is against our current Code Owner guidelines.

TylerHelmuth committed:
PR feedback

TylerHelmuth committed:
Check for empty owners

TylerHelmuth commented:
@codeboten addressed feedback, but further testing showed that trying to assign a user that doesn't have permission to be assigned didn't fail like I wanted it to (but pinging worked).  What behavior do we want?

TylerHelmuth commented:
Strange, I didn't see that error when testing on my fork. I will investigate tomorrow. 

@bogdandrutu are there any special configuration rules this repo uses for actions?

TylerHelmuth committed:
Update permissions

TylerHelmuth committed:
update permissions

TylerHelmuth committed:
Update permissions

TylerHelmuth committed:
Update permissions

TylerHelmuth commented:
@bogdandrutu the issue is definitely the fact that the GITHUB_TOKEN has read permissions. 

![Screen Shot 2022-09-21 at 9 54 44 PM](https://user-images.githubusercontent.com/12352919/191654793-eeaa5bfb-fe35-4f24-b0e5-590ad055fb83.png)

In the run in my fork the token had write permission.  I have attempted to update the permissions using the job's `permissions` section, but they are not being honored.  

I suspect, although I can't prove yet, that the issue is that this workflow isn't officially part of the repo yet. I think GH is attempting to keep things secure by not allowing workflows introduced by PRs to ever have write permissions.

TylerHelmuth commented:
I _think_ this is the proof: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

![Screen Shot 2022-09-21 at 10 06 13 PM](https://user-images.githubusercontent.com/12352919/191656161-21b68db8-9d05-46ed-be4e-d4f536781054.png)

TylerHelmuth committed:
Merge branch 'open-telemetry:main' into auto-assign-codeowners

bogdandrutu commented on .github/workflows/assign-codeowners.yml:
```diff
@@ -0,0 +1,23 @@
+name: 'Assign code owners'
+on:
+  pull_request:
```
See https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/.github/workflows/auto-assign-owners.yml#L3, I have not thought about any security issues, so please do that :)

TylerHelmuth replied:
Yea that'll definitely give PRs from forks write access. There are definitely some security risks according to GitHub (I'll post the article tomorrow).

I'll do some experimenting to see if that target is necessary still once the workflow is checked in, but I don't think it is.

TylerHelmuth replied:
Here is the article talking about pull_request_target: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/.  Best practice with this event is to not checkout code.

bogdandrutu commented:
@TylerHelmuth this requires manual work to assign a label. I think we should look into something like prow bot, but that was said that is too much, but we can at least use https://github.com/marketplace/actions/prow-github-actions which does more automation. And we start having the "OWNERS" files, and people get automatically assign (I believe) without having to add a label. I would prefer to not have to add a label and also to get rid of the auto-assign that we have right now.

TylerHelmuth commented:
@bogdandrutu I reviewed that action before submitting this PR and unless I'm missing something it does not support automatically assigning owners when the PR is create; you must add a `/assign` comment to the issueand you must supply the users being assigned.

It looks like it does support automatic labeling, though, which could be used in conjunction with this PR to automatically assign code owners.  But then we'd have to manage a `labels.yaml` with all the component labels.

TylerHelmuth committed:
Add auto-labeler

TylerHelmuth committed:
rename action

TylerHelmuth commented:
I think combining labeler and assign-code-owners is possible, but its gonna take some extra work since I can't checkout code.  @bogdandrutu @codeboten @open-telemetry/collector-contrib-approvers how do you want to proceed?
